generation.go

```diff
@@ -0,0 +1,388 @@
+					}
+
+					if strings.HasPrefix(f.CName, "clang_create") || strings.HasPrefix(f.CName, "clang_get") || strings.HasSuffix(f.CName, "_create") {
+						fname = "New" + fname
+					}
+
+					added = g.addMethod(f, fname, "", rtc)
+				}
+				if !added {
+					f.Name = UpperFirstCharacter(f.Name)
+
+					clangFile.Functions = append(clangFile.Functions, f)
+
+					added = true
+				}
+			}
+		}
+
+		if !added {
+			fmt.Println("Unused function:", f.Name)
+		}
+	}
+
+	for _, e := range g.enums {
+		if err := e.addEnumStringMethods(); err != nil {
+			return fmt.Errorf("Cannot generate enum string methods: %v", err)
+		}
+
+		for i, m := range e.Methods {
+			if r := g.handleMethod(e.Name, m); r != "" {
+				e.Methods[i] = r
+			}
+			switch m := m.(type) {
+			case *Function:
+				if len(m.Parameters) > 0 && !m.Parameters[0].Type.IsSlice && m.Parameters[0].Type.GoName == e.Name {
+					m.Receiver = Receiver{
+						Name: commonReceiverName(e.Name),
+						Type: m.Parameters[0].Type,
+					}
+
+					g.setIsPointerComposition(&m.Receiver.Type)
+				}
+				for i := range m.Parameters {
+					g.setIsPointerComposition(&m.Parameters[i].Type)
+				}
+
+				e.IncludeFiles.unifyIncludeFiles(m.IncludeFiles)
+
+				e.Methods[i] = m.generate()
+			}
+		}
+
+		if err := e.generate(); err != nil {
+			return fmt.Errorf("Cannot generate enum: %v", err)
+		}
+	}
+
+	for _, s := range g.structs {
+		if err := s.addMemberGetters(); err != nil {
+			return fmt.Errorf("Cannot generate struct member getters: %v", err)
+		}
+
+		for i, m := range s.Methods {
+			if r := g.handleMethod(s.Name, m); r != "" {
+				s.Methods[i] = r
+			}
+			switch m := m.(type) {
+			case *Function:
+				s.IncludeFiles.unifyIncludeFiles(m.IncludeFiles)
+			}
+		}
+
+		if err := s.generate(); err != nil {
+			return fmt.Errorf("Cannot generate struct: %v", err)
+		}
+	}
+
+	if len(clangFile.Functions) > 0 {
+		for _, m := range clangFile.Functions {
+			switch m := m.(type) {
+			case *Function:
+				clangFile.IncludeFiles.unifyIncludeFiles(m.IncludeFiles)
+			}
+		}
+
+		for i, m := range clangFile.Functions {
+			if r := g.handleMethod("", m); r != "" {
+				clangFile.Functions[i] = r
+			}
+		}
+
+		if err := clangFile.generate(); err != nil {
+			return fmt.Errorf("Cannot generate clang file: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (g *Generation) addMethod(f *Function, fname string, fnamePrefix string, rt Receiver) bool {
+	fname = UpperFirstCharacter(fnamePrefix + UpperFirstCharacter(fname))
+
+	if e, ok := g.lookupEnum[rt.Type.GoName]; ok {
+		f.Name = fname
+
+		e.Methods = append(e.Methods, f)
+
+		return true
+	} else if s, ok := g.lookupStruct[rt.Type.GoName]; ok && s.CName != "CXString" {
+		f.Name = fname
+
+		if !rt.Type.IsSlice && rt.Type.PointerLevel > 0 {
+			s.IsPointerComposition = true
+		}
+
+		s.Methods = append(s.Methods, f)
+
+		return true
+	}
+
+	return false
+}
+
+func (g *Generation) handleMethod(receiverName string, m interface{}) string {
```

Comment:
Have a look at where "handleMethod" is called. The all do something like this
```go
if r := h.handleMethod("", m); r != "" {
 clangFile.Functions[i] = r
}
```

That is because the method is of the type `{}interface` because AFAIK we can have an object of type `string` and type `*Function` in there. We could add a case for the type `string` in handleMethod which could then just return the value of the method, and then we can reduce the calls to something like this
```go
clangFile.Functions[i] = h.handleMethod("", m)
```